src/cluster.c

```diff
@@ -4948,6 +4962,163 @@ void addNodeReplyForClusterSlot(client *c, clusterNode *node, int start_slot, in
+     * 5. endpoint
+     * 5. role
+     * 6. replication offset (in case of replica only).
+     * 7. health
+     */
+    int reply_cnt = 0;
+    void *node_replylen = addReplyDeferredLen(c);
+    addReplyBulkCString(c, "id");
+    addReplyBulkCBuffer(c, node->name, CLUSTER_NAMELEN);
+    reply_cnt++;
+    addReplyBulkCString(c, "port");
+    addReplyLongLong(c, node->port);
+    reply_cnt++;
+    addReplyBulkCString(c, "ip");
+    addReplyBulkCString(c, node->ip);
+    reply_cnt++;
+    if (node->hostname) {
+        addReplyBulkCString(c, "hostname");
+        addReplyBulkCString(c, node->hostname);
+        reply_cnt++;
+    }
+    const char *endpoint = getPreferredEndpoint(node);
+    if (!strcmp(endpoint, "")) {
+        addReplyBulkCString(c, "endpoint");
+        addReplyBulkCString(c, endpoint);
+        reply_cnt++;
+    }
+
+    addReplyBulkCString(c, "role");
+    if (nodeIsSlave(node)) {
+        addReplyBulkCString(c, "replica");
+        addReplyBulkCString(c, "replication-offset");
+        addReplyLongLong(c, node->repl_offset);
+        reply_cnt+=2;
+    } else {
+        addReplyBulkCString(c, "primary");
+        reply_cnt++;
+    }
+
+    addReplyBulkCString(c, "health");
+    const char *health_msg = NULL;
+    if (nodeFailed(node)) {
+        health_msg = "FAIL";
+    } else if (nodeIsSlave(node) && node->repl_offset == 0) {
+        health_msg = "LOADING";
+    } else {
+        health_msg = "ONLINE";
+    }
+    addReplyBulkCString(c, health_msg);
+    reply_cnt++;
+
+    setDeferredMapLen(c, node_replylen, reply_cnt);
+}
+
+void addNodeReplyForClusterShard(client *c, clusterNode *node, list *slotInfo) {
+    addReplyMapLen(c, 2);
+    addReplyBulkCString(c,"slots");
+    uint16_t slotInfoCount = 0;
+    if (slotInfo != NULL) {
+        slotInfoCount = listLength(slotInfo);
+        addReplyArrayLen(c, slotInfoCount);
+        listIter li;
+        listRewind(slotInfo, &li);
+        listNode *ln;
+        while((ln = listNext(&li))) {
+            addReplyBulkSds(c, listNodeValue(ln));
+        }
+    } else {
+        addReplyArrayLen(c, slotInfoCount);
+    }
+    addReplyBulkCString(c,"nodes");
+    list *nodes_for_slot = clusterGetNodesServingMySlots(node);
+    if (nodes_for_slot != NULL) {
+        addReplyArrayLen(c, listLength(nodes_for_slot));
+        if (listLength(nodes_for_slot) != 0) {
+            listIter li;
+            listNode *ln;
+            listRewind(nodes_for_slot, &li);
+            while ((ln = listNext(&li))) {
+                clusterNode *node = listNodeValue(ln);
+                addNodeDetailsToShardReply(c, node);
+            }
+            listRelease(nodes_for_slot);
+        }
+    } else {
+        addReplyArrayLen(c, 0);
+    }
+}
+
+
+void clusterReplyShards(client *c) {
+    clusterNode *n = NULL;
+    int start = -1;
+    void *shard_replylen = addReplyDeferredLen(c);
+    int shard_cnt = 0;
+    dict *nodeToSlotPair = dictCreate(&clusterNodesToSlotDictType);
```

Comment:
Creating a temporary dict seems unnecessary. Maybe we can avoid it?

I think it's cheaper to iterate over the structures and add replies directly without allocating any extra structures, even if we might need to iterate over the nodes one extra time e.g. once just to count the shards to `addReplyArrayLen(c, numshards);`.

If we need to create a structure like this one, then perhaps it can be stored permanently or maybe we can change the existing structures to be more optimized for CLUSTER SHARDS (if possible).

Reply:
I think we could re-use the iterator logic for `CLUSTER NODES` that loops through all the slots an caches the response on each node. We can then re-iterate through the cached nodes to add the output.